// @sap/ux-specification/dist/src/sync/v4/application.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../common/decoration/decorators");
exports.getSettingsPathForPage = (pageName) => {
    return 'sap.ui5/routing/targets/' + pageName[0] + '/options/settings';
};
exports.getControlConfigurationPath = (pageName) => {
    return exports.getSettingsPathForPage(pageName) + '/controlConfiguration';
};
exports.getTableSettingsPathForLR = (pageName) => {
    return exports.getControlConfigurationPath(pageName) + '/@com.sap.vocabularies.UI.v1.LineItem/tableSettings';
};
exports.getTableSettingsPathForOP = (pageName, sectionId) => {
    return exports.getControlConfigurationPath(pageName) + '/' + sectionId + '/tableSettings';
};
exports.getCustomSectionPath = (pageName, ...pathParts) => {
    return exports.getSettingsPathForPage(pageName) + '/content/body/sections/' + pathParts.join('/');
};
class ApplicationV4 {
}
__decorate([
    decorators_1.childId('settings')
], ApplicationV4.prototype, "settings", void 0);
exports.ApplicationV4 = ApplicationV4;
class AppSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "title", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "description", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.ui5/routing/config'
        }
    }),
    decorators_1.validity({
        since: '1.78.0'
    })
], AppSettings.prototype, "flexibleColumnLayout", void 0);
exports.AppSettings = AppSettings;
