// @sap/ux-specification/dist/src/sync/v4/export/pages/ObjectPage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
let ObjectPage = class ObjectPage {
};
__decorate([
    decorators_1.childId('header')
], ObjectPage.prototype, "header", void 0);
__decorate([
    decorators_1.childId('layout')
], ObjectPage.prototype, "layout", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet)
        }
    }),
    decorators_1.validity({
        since: '1.74.0'
    })
], ObjectPage.prototype, "variantManagement", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet)
        }
    }),
    decorators_1.validity({
        since: '1.74.0'
    })
], ObjectPage.prototype, "initialLoad", void 0);
__decorate([
    decorators_1.childId('sections')
], ObjectPage.prototype, "sections", void 0);
ObjectPage = __decorate([
    decorators_1.pageLayoutInformation({
        id: 'sap.suite.ui.generic.template.ObjectPage.view.Details',
        pageId: 'sap.suite.ui.generic.template.ObjectPage'
    })
], ObjectPage);
exports.ObjectPage = ObjectPage;
