// @sap/ux-specification/dist/src/sync/v2/export/controls/Card.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "entitySet", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "title", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "subTitle", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "identificationAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "selectionAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CommonCardSettings.prototype, "presentationAnnotationPath", void 0);
exports.CommonCardSettings = CommonCardSettings;
class CustomCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CustomCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CustomCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], CustomCard.prototype, "settings", void 0);
exports.CustomCard = CustomCard;
class ListCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "settings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "listCardSettings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "tabs", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "listFlavor", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "listType", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "annotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "sortBy", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "sortOrder", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], ListCard.prototype, "addODataSelect", void 0);
exports.ListCard = ListCard;
class AnalyticalCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "chartType", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "settings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "analyticalCardSettings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "chartAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "dataPointAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], AnalyticalCard.prototype, "tabs", void 0);
exports.AnalyticalCard = AnalyticalCard;
class StackCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "settings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "stackCardSettings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "tabs", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "itemText", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], StackCard.prototype, "objectStreamCardsNavigationProperty", void 0);
exports.StackCard = StackCard;
class LinklistCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "settings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "linkListCardSettings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "entitySet", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "title", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "subTitle", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "targetUri", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "listFlavor", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "sortBy", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "sortOrder", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "headerAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "identificationAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "defaultSpan", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "staticContent", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], LinklistCard.prototype, "tabs", void 0);
exports.LinklistCard = LinklistCard;
class TableCard extends CommonCardSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "model", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "template", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "disableTableCardFlexibility", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "settings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "tableCardSettings", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "annotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "listFlavor", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "listType", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "dataPointAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "selectionPresentationAnnotationPath", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (cardKeys) => application_1.getCardSettingsPath(cardKeys)
        }
    })
], TableCard.prototype, "tabs", void 0);
exports.TableCard = TableCard;
