// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageTable.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
const application_1 = require("../../application");
const common_1 = require("../../../common");
const getSmartTableControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, title) => {
    const breadCrumbsCopy = breadcrumbs.slice(1, 2);
    let suffixId;
    if (controlType() === 'sap.ui.comp.smarttable.SmartTable') {
        suffixId = 'Table';
    }
    else if (controlType() === 'sap.m.Table') {
        suffixId = 'responsiveTable';
    }
    breadCrumbsCopy.push(suffixId);
    breadCrumbsCopy[0] = decoration_1.convertSectionId(breadCrumbsCopy[0], title);
    return decoration_1.buildControlIdFromParent(baseId, breadCrumbsCopy);
};
class ObjectPageCommonTableSettings {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => common_1.ControlType.SmartTable,
            exportFunction: decoration_1.exportToFlexChange
        }
    }),
    decorators_1.validity({
        since: '1.28.0'
    })
], ObjectPageCommonTableSettings.prototype, "editable", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.54.0'
    })
], ObjectPageCommonTableSettings.prototype, "multiSelect", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.70.0'
    })
], ObjectPageCommonTableSettings.prototype, "selectAll", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.70.0'
    })
], ObjectPageCommonTableSettings.prototype, "selectionLimit", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.63.0'
    })
], ObjectPageCommonTableSettings.prototype, "quickVariantSelection", void 0);
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => common_1.ControlType.SmartTable,
            exportFunction: decoration_1.exportToFlexChange
        }
    }),
    decorators_1.validity({
        since: '1.26.0'
    })
], ObjectPageCommonTableSettings.prototype, "useExportToExcel", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.48.5'
    })
], ObjectPageCommonTableSettings.prototype, "createMode", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.61.0'
    })
], ObjectPageCommonTableSettings.prototype, "disableDefaultInlineCreateSort", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.78.0'
    })
], ObjectPageCommonTableSettings.prototype, "variantManagement", void 0);
exports.ObjectPageCommonTableSettings = ObjectPageCommonTableSettings;
class ObjectPageResponsiveTable extends ObjectPageCommonTableSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    })
], ObjectPageResponsiveTable.prototype, "type", void 0);
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => common_1.ControlType.Table,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], ObjectPageResponsiveTable.prototype, "includeItemInSelection", void 0);
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => common_1.ControlType.Table,
            exportFunction: decoration_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnum
    }),
    decorators_1.validity({
        since: '1.52.0'
    })
], ObjectPageResponsiveTable.prototype, "popinLayout", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageResponsiveTable.prototype, "inlineDelete", void 0);
exports.ObjectPageResponsiveTable = ObjectPageResponsiveTable;
class ObjectPageAnalyticalTable extends ObjectPageCommonTableSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    })
], ObjectPageAnalyticalTable.prototype, "type", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.67.0'
    })
], ObjectPageAnalyticalTable.prototype, "condensedTableLayout", void 0);
exports.ObjectPageAnalyticalTable = ObjectPageAnalyticalTable;
class ObjectPageGridTable extends ObjectPageCommonTableSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    })
], ObjectPageGridTable.prototype, "type", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.67.0'
    })
], ObjectPageGridTable.prototype, "condensedTableLayout", void 0);
exports.ObjectPageGridTable = ObjectPageGridTable;
class ObjectPageTreeTable extends ObjectPageCommonTableSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getTableSettingsOBJsectionPathV2(pageKeys, sectionId)
        }
    })
], ObjectPageTreeTable.prototype, "type", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageKeys, sectionId) => application_1.getAppSettingOBJsectionPathV2(pageKeys, sectionId)
        }
    }),
    decorators_1.validity({
        since: '1.67.0'
    })
], ObjectPageTreeTable.prototype, "condensedTableLayout", void 0);
exports.ObjectPageTreeTable = ObjectPageTreeTable;
