// @sap/ux-specification/dist/src/specification/common/page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PageType;
(function (PageType) {
    PageType["ObjectPage"] = "ObjectPage";
    PageType["ListReport"] = "ListReport";
    PageType["OverviewPage"] = "OverviewPage";
    PageType["CustomPage"] = "CustomPage";
    PageType["AnalyticalListPage"] = "AnalyticalListPage";
})(PageType = exports.PageType || (exports.PageType = {}));
var PageTypeV4;
(function (PageTypeV4) {
    PageTypeV4["ObjectPage"] = "ObjectPage";
    PageTypeV4["ListReport"] = "ListReport";
    PageTypeV4["CustomPage"] = "CustomPage";
    PageTypeV4["AnalyticalListPage"] = "AnalyticalListPage";
})(PageTypeV4 = exports.PageTypeV4 || (exports.PageTypeV4 = {}));
var FioriElementsVersion;
(function (FioriElementsVersion) {
    FioriElementsVersion["v2"] = "v2";
    FioriElementsVersion["v4"] = "v4";
})(FioriElementsVersion = exports.FioriElementsVersion || (exports.FioriElementsVersion = {}));
var OdataVersion;
(function (OdataVersion) {
    OdataVersion["v2"] = "v2";
    OdataVersion["v4"] = "v4";
})(OdataVersion = exports.OdataVersion || (exports.OdataVersion = {}));
