// @sap/ux-specification/dist/src/specification/common/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlexibleColumnLayoutType;
(function (FlexibleColumnLayoutType) {
    FlexibleColumnLayoutType["OneColumn"] = "OneColumn";
    FlexibleColumnLayoutType["TwoColumnsBeginExpanded"] = "TwoColumnsBeginExpanded";
    FlexibleColumnLayoutType["TwoColumnsMidExpanded"] = "TwoColumnsMidExpanded";
    FlexibleColumnLayoutType["MidColumnFullScreen"] = "MidColumnFullScreen";
    FlexibleColumnLayoutType["ThreeColumnsMidExpanded"] = "ThreeColumnsMidExpanded";
    FlexibleColumnLayoutType["ThreeColumnsEndExpanded"] = "ThreeColumnsEndExpanded";
    FlexibleColumnLayoutType["ThreeColumnsMidExpandedEndHidden"] = "ThreeColumnsMidExpandedEndHidden";
    FlexibleColumnLayoutType["ThreeColumnsBeginExpandedEndHidden"] = "ThreeColumnsBeginExpandedEndHidden";
    FlexibleColumnLayoutType["EndColumnFullScreen"] = "EndColumnFullScreen";
})(FlexibleColumnLayoutType = exports.FlexibleColumnLayoutType || (exports.FlexibleColumnLayoutType = {}));
