// @sap/ux-specification/dist/src/specification/v4/controls/Table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SelectionMode;
(function (SelectionMode) {
    SelectionMode["Multi"] = "Multi";
    SelectionMode["None"] = "None";
    SelectionMode["Single"] = "Single";
    SelectionMode["Auto"] = "Auto";
})(SelectionMode = exports.SelectionMode || (exports.SelectionMode = {}));
var TableTypeV4;
(function (TableTypeV4) {
    TableTypeV4["ResponsiveTable"] = "ResponsiveTable";
    TableTypeV4["GridTable"] = "GridTable";
    /* Not supported yet:
    AnalyticalTable = "AnalyticalTable",
    TreeTable = "TreeTable"*/
})(TableTypeV4 = exports.TableTypeV4 || (exports.TableTypeV4 = {}));
var TableCreationModeType;
(function (TableCreationModeType) {
    TableCreationModeType["NewPage"] = "NewPage";
    TableCreationModeType["Inline"] = "Inline";
    TableCreationModeType["CreationRow"] = "CreationRow";
})(TableCreationModeType = exports.TableCreationModeType || (exports.TableCreationModeType = {}));
var VariantManagementTypeListReport;
(function (VariantManagementTypeListReport) {
    VariantManagementTypeListReport["None"] = "None";
    VariantManagementTypeListReport["Control"] = "Control";
    VariantManagementTypeListReport["Page"] = "Page";
})(VariantManagementTypeListReport = exports.VariantManagementTypeListReport || (exports.VariantManagementTypeListReport = {}));
