// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageCustomSection.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const v4_1 = require("../../../../specification/v4");
const common_1 = require("../../../../specification/common");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ObjectPageCustomSectionFragment {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet, sectionId) => application_1.getCustomSectionPath(entitySet, sectionId, 'position'),
            key: 'anchor'
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifest, sectionId) => {
                // Id is read from path's section - it is object's key
                return sectionId;
            },
            export: false
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "id", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet, sectionId) => application_1.getCustomSectionPath(entitySet, sectionId, 'position'),
            key: 'placement'
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getCustomSectionPath
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "title", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifest) => {
                // Currently only XML supported, but idea is that we need to check 'manifest.type' and allow other types - maybe JS, JSON, etc.
                if (manifest.type.startsWith(v4_1.CustomSectionViewTypesV4.XML)) {
                    return v4_1.CustomSectionViewTypesV4.XML;
                }
                return '';
            },
            export: (manifest, object) => {
                // Currently only 'XMLFragment', but in future we can add logic here
                let type = '';
                if ('type' in object) {
                    type += object.type;
                }
                if ('className' in object) {
                    if (object.className === common_1.SAPUI5_FRAGMENT_CLASS) {
                        type += 'Fragment';
                    }
                    else {
                        type += object.className;
                    }
                }
                manifest.type = type;
            }
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "type", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet, sectionId) => application_1.getCustomSectionPath(entitySet, sectionId),
            import: (manifest) => {
                // Currently only XML Fragment supported, but idea is that we need to check 'value.type'
                // f.e. type === 'XMLFragment' ? 'sap.ui.core.Fragment' : ...
                if (manifest.type.startsWith(v4_1.CustomSectionViewTypesV4.XML)) {
                    if (manifest.type === 'XMLFragment') {
                        return common_1.SAPUI5_FRAGMENT_CLASS;
                    }
                    else {
                        return manifest.type.slice(v4_1.CustomSectionViewTypesV4.XML.length);
                    }
                }
                // Fallback - if there is unsupported value
                return manifest.type;
            },
            export: false
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "className", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getCustomSectionPath,
            key: 'name'
        }
    }),
    decorators_1.validity({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "fragmentName", void 0);
exports.ObjectPageCustomSectionFragment = ObjectPageCustomSectionFragment;
