// @sap-ux/vocabularies-types/types/generated/Analytics_Edm.ts
import * as com_sap_vocabularies_Analytics_v1 from "./Analytics";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
	"DrillURL"?: com_sap_vocabularies_Analytics_v1.DrillURL;
	"AggregatedProperties"?: com_sap_vocabularies_Analytics_v1.AggregatedProperties;
}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Analytics {
	"Dimension"?: com_sap_vocabularies_Analytics_v1.Dimension;
	"Measure"?: com_sap_vocabularies_Analytics_v1.Measure;
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
	"PlanningAction"?: com_sap_vocabularies_Analytics_v1.PlanningAction;
}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Analytics {
	"AccumulativeMeasure"?: com_sap_vocabularies_Analytics_v1.AccumulativeMeasure;
	"RolledUpPropertyCount"?: com_sap_vocabularies_Analytics_v1.RolledUpPropertyCount;
}
