// @sap-ux/vocabularies-types/types/generated/Auth_Edm.ts
import * as Org_OData_Authorization_V1 from "./Auth";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Auth {
	"SecuritySchemes"?: Org_OData_Authorization_V1.SecuritySchemes;
	"Authorizations"?: Org_OData_Authorization_V1.Authorizations;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Auth {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Auth {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Auth {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Auth {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Auth {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Auth {}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Auth {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Auth {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Auth {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Auth {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Auth {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Auth {}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Auth {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Auth {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Auth {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Auth {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Auth {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Auth {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Auth {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Auth {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Auth {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Auth {}
