// @sap-ux/vocabularies-types/types/generated/Capabilities_Edm.ts
import * as Org_OData_Capabilities_V1 from "./Capabilities";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Capabilities {
	"ConformanceLevel"?: Org_OData_Capabilities_V1.ConformanceLevel;
	"SupportedFormats"?: Org_OData_Capabilities_V1.SupportedFormats;
	"SupportedMetadataFormats"?: Org_OData_Capabilities_V1.SupportedMetadataFormats;
	"AcceptableEncodings"?: Org_OData_Capabilities_V1.AcceptableEncodings;
	"AsynchronousRequestsSupported"?: Org_OData_Capabilities_V1.AsynchronousRequestsSupported;
	"BatchContinueOnErrorSupported"?: Org_OData_Capabilities_V1.BatchContinueOnErrorSupported;
	"IsolationSupported"?: Org_OData_Capabilities_V1.IsolationSupported;
	"CrossJoinSupported"?: Org_OData_Capabilities_V1.CrossJoinSupported;
	"CallbackSupported"?: Org_OData_Capabilities_V1.CallbackSupported;
	"SelectSupport"?: Org_OData_Capabilities_V1.SelectSupport;
	"BatchSupported"?: Org_OData_Capabilities_V1.BatchSupported;
	"BatchSupport"?: Org_OData_Capabilities_V1.BatchSupport;
	"FilterFunctions"?: Org_OData_Capabilities_V1.FilterFunctions;
	"KeyAsSegmentSupported"?: Org_OData_Capabilities_V1.KeyAsSegmentSupported;
	"QuerySegmentSupported"?: Org_OData_Capabilities_V1.QuerySegmentSupported;
	"DeepInsertSupport"?: Org_OData_Capabilities_V1.DeepInsertSupport;
	"DeepUpdateSupport"?: Org_OData_Capabilities_V1.DeepUpdateSupport;
	"AnnotationValuesInQuerySupported"?: Org_OData_Capabilities_V1.AnnotationValuesInQuerySupported;
	"ModificationQueryOptions"?: Org_OData_Capabilities_V1.ModificationQueryOptions;
	"CustomHeaders"?: Org_OData_Capabilities_V1.CustomHeaders;
	"CustomQueryOptions"?: Org_OData_Capabilities_V1.CustomQueryOptions;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Capabilities {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Capabilities {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Capabilities {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Capabilities {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Capabilities {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Capabilities {
	"MediaLocationUpdateSupported"?: Org_OData_Capabilities_V1.MediaLocationUpdateSupported;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Capabilities {
	"ChangeTracking"?: Org_OData_Capabilities_V1.ChangeTracking;
}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Capabilities {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Capabilities {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Capabilities {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Capabilities {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Capabilities {
	"CallbackSupported"?: Org_OData_Capabilities_V1.CallbackSupported;
	"ChangeTracking"?: Org_OData_Capabilities_V1.ChangeTracking;
	"CountRestrictions"?: Org_OData_Capabilities_V1.CountRestrictions;
	"NavigationRestrictions"?: Org_OData_Capabilities_V1.NavigationRestrictions;
	"IndexableByKey"?: Org_OData_Capabilities_V1.IndexableByKey;
	"TopSupported"?: Org_OData_Capabilities_V1.TopSupported;
	"SkipSupported"?: Org_OData_Capabilities_V1.SkipSupported;
	"ComputeSupported"?: Org_OData_Capabilities_V1.ComputeSupported;
	"SelectSupport"?: Org_OData_Capabilities_V1.SelectSupport;
	"FilterFunctions"?: Org_OData_Capabilities_V1.FilterFunctions;
	"FilterRestrictions"?: Org_OData_Capabilities_V1.FilterRestrictions;
	"SortRestrictions"?: Org_OData_Capabilities_V1.SortRestrictions;
	"ExpandRestrictions"?: Org_OData_Capabilities_V1.ExpandRestrictions;
	"SearchRestrictions"?: Org_OData_Capabilities_V1.SearchRestrictions;
	"InsertRestrictions"?: Org_OData_Capabilities_V1.InsertRestrictions;
	"DeepInsertSupport"?: Org_OData_Capabilities_V1.DeepInsertSupport;
	"UpdateRestrictions"?: Org_OData_Capabilities_V1.UpdateRestrictions;
	"DeepUpdateSupport"?: Org_OData_Capabilities_V1.DeepUpdateSupport;
	"DeleteRestrictions"?: Org_OData_Capabilities_V1.DeleteRestrictions;
	"CollectionPropertyRestrictions"?: Org_OData_Capabilities_V1.CollectionPropertyRestrictions;
	"ReadRestrictions"?: Org_OData_Capabilities_V1.ReadRestrictions;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Capabilities {
	"ChangeTracking"?: Org_OData_Capabilities_V1.ChangeTracking;
	"CountRestrictions"?: Org_OData_Capabilities_V1.CountRestrictions;
	"NavigationRestrictions"?: Org_OData_Capabilities_V1.NavigationRestrictions;
	"SelectSupport"?: Org_OData_Capabilities_V1.SelectSupport;
	"ExpandRestrictions"?: Org_OData_Capabilities_V1.ExpandRestrictions;
	"UpdateRestrictions"?: Org_OData_Capabilities_V1.UpdateRestrictions;
	"DeleteRestrictions"?: Org_OData_Capabilities_V1.DeleteRestrictions;
	"CollectionPropertyRestrictions"?: Org_OData_Capabilities_V1.CollectionPropertyRestrictions;
	"ReadRestrictions"?: Org_OData_Capabilities_V1.ReadRestrictions;
}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Capabilities {
	"ModificationQueryOptions"?: Org_OData_Capabilities_V1.ModificationQueryOptions;
}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Capabilities {
	"ChangeTracking"?: Org_OData_Capabilities_V1.ChangeTracking;
}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Capabilities {
	"OperationRestrictions"?: Org_OData_Capabilities_V1.OperationRestrictions;
	"ModificationQueryOptions"?: Org_OData_Capabilities_V1.ModificationQueryOptions;
}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Capabilities {
	"ChangeTracking"?: Org_OData_Capabilities_V1.ChangeTracking;
	"OperationRestrictions"?: Org_OData_Capabilities_V1.OperationRestrictions;
}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Capabilities {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Capabilities {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Capabilities {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Capabilities {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Capabilities {}
