// @sap-ux/vocabularies-types/types/generated/CodeList.ts
import * as Core from "./Core";
import * as Common from "./Common";
import * as Edm from "../Edm";
import AnnotationTerm = Edm.AnnotationTerm;
import PropertyAnnotationValue = Edm.PropertyAnnotationValue;
import EnumValue = Edm.EnumValue;
import ComplexType = Edm.ComplexType;

// Term
/**
  An entity set containing the code list for currencies
*/
export type CurrencyCodes = AnnotationTerm<CodeListSourceTypes>;

// Term
/**
  An entity set containing the code list for units of measure
*/
export type UnitsOfMeasure = AnnotationTerm<CodeListSourceTypes>;

// ComplexType
/**
  An entity set containing the code list for currencies
*/
export type CodeListSource = ComplexType & {
	$Type: CodeListAnnotationTypes.CodeListSource;
	/**
        URL of a CSDL document describing an entity set for a code list
    */
	Url: PropertyAnnotationValue<Edm.String>;
	/**
        Name of the entity set for the code list
    */
	CollectionPath: PropertyAnnotationValue<Edm.String>;
};

// Term
/**
  Property containing standard code values
*/
export type StandardCode = AnnotationTerm<Edm.PropertyPath>;

// Term
/**
  Property containing code values that can be used for visualization
*/
export type ExternalCode = AnnotationTerm<Edm.PropertyPath>;

// Term
/**
  Property contains a Configuration Deprecation Code
*/
export type IsConfigurationDeprecationCode = AnnotationTerm<Edm.Boolean>;
export const enum CodeListAnnotationTerms {
	CurrencyCodes = "com.sap.vocabularies.CodeList.v1.CurrencyCodes",
	UnitsOfMeasure = "com.sap.vocabularies.CodeList.v1.UnitsOfMeasure",
	StandardCode = "com.sap.vocabularies.CodeList.v1.StandardCode",
	ExternalCode = "com.sap.vocabularies.CodeList.v1.ExternalCode",
	IsConfigurationDeprecationCode = "com.sap.vocabularies.CodeList.v1.IsConfigurationDeprecationCode"
}
export const enum CodeListAnnotationTypes {
	CodeListSource = "com.sap.vocabularies.CodeList.v1.CodeListSource"
}

export type CodeListSourceTypes = AnnotationTerm<CodeListSource>;
