// @sap-ux/vocabularies-types/types/generated/CodeList_Edm.ts
import * as com_sap_vocabularies_CodeList_v1 from "./CodeList";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_CodeList {
	"CurrencyCodes"?: com_sap_vocabularies_CodeList_v1.CurrencyCodes;
	"UnitsOfMeasure"?: com_sap_vocabularies_CodeList_v1.UnitsOfMeasure;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_CodeList {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_CodeList {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_CodeList {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_CodeList {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_CodeList {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_CodeList {
	"StandardCode"?: com_sap_vocabularies_CodeList_v1.StandardCode;
	"ExternalCode"?: com_sap_vocabularies_CodeList_v1.ExternalCode;
	"IsConfigurationDeprecationCode"?: com_sap_vocabularies_CodeList_v1.IsConfigurationDeprecationCode;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_CodeList {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_CodeList {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_CodeList {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_CodeList {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_CodeList {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_CodeList {}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_CodeList {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_CodeList {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_CodeList {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_CodeList {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_CodeList {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_CodeList {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_CodeList {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_CodeList {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_CodeList {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_CodeList {}
