// @sap-ux/vocabularies-types/types/generated/Core_Edm.ts
import * as Org_OData_Core_V1 from "./Core";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Core {
	"ODataVersions"?: Org_OData_Core_V1.ODataVersions;
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"DereferenceableIDs"?: Org_OData_Core_V1.DereferenceableIDs;
	"ConventionalIDs"?: Org_OData_Core_V1.ConventionalIDs;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Core {
	"SchemaVersion"?: Org_OData_Core_V1.SchemaVersion;
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"DefaultNamespace"?: Org_OData_Core_V1.DefaultNamespace;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Core {
	"SchemaVersion"?: Org_OData_Core_V1.SchemaVersion;
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"AcceptableMediaTypes"?: Org_OData_Core_V1.AcceptableMediaTypes;
	"AdditionalProperties"?: Org_OData_Core_V1.AdditionalProperties;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"AlternateKeys"?: Org_OData_Core_V1.AlternateKeys;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"AdditionalProperties"?: Org_OData_Core_V1.AdditionalProperties;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"IsLanguageDependent"?: Org_OData_Core_V1.IsLanguageDependent;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"Immutable"?: Org_OData_Core_V1.Immutable;
	"Computed"?: Org_OData_Core_V1.Computed;
	"ComputedDefaultValue"?: Org_OData_Core_V1.ComputedDefaultValue;
	"IsURL"?: Org_OData_Core_V1.IsURL;
	"AcceptableMediaTypes"?: Org_OData_Core_V1.AcceptableMediaTypes;
	"MediaType"?: Org_OData_Core_V1.MediaType;
	"IsMediaType"?: Org_OData_Core_V1.IsMediaType;
	"ContentDisposition"?: Org_OData_Core_V1.ContentDisposition;
	"AutoExpand"?: Org_OData_Core_V1.AutoExpand;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"Ordered"?: Org_OData_Core_V1.Ordered;
	"PositionalInsert"?: Org_OData_Core_V1.PositionalInsert;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"AutoExpand"?: Org_OData_Core_V1.AutoExpand;
	"AutoExpandReferences"?: Org_OData_Core_V1.AutoExpandReferences;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"Ordered"?: Org_OData_Core_V1.Ordered;
	"PositionalInsert"?: Org_OData_Core_V1.PositionalInsert;
	"AlternateKeys"?: Org_OData_Core_V1.AlternateKeys;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"IsLanguageDependent"?: Org_OData_Core_V1.IsLanguageDependent;
	"RequiresType"?: Org_OData_Core_V1.RequiresType;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"IsURL"?: Org_OData_Core_V1.IsURL;
	"AcceptableMediaTypes"?: Org_OData_Core_V1.AcceptableMediaTypes;
	"MediaType"?: Org_OData_Core_V1.MediaType;
	"IsMediaType"?: Org_OData_Core_V1.IsMediaType;
	"ContentDisposition"?: Org_OData_Core_V1.ContentDisposition;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"Ordered"?: Org_OData_Core_V1.Ordered;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"OptionalParameter"?: Org_OData_Core_V1.OptionalParameter;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Example"?: Org_OData_Core_V1.Example;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"Ordered"?: Org_OData_Core_V1.Ordered;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ResourcePath"?: Org_OData_Core_V1.ResourcePath;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"OptimisticConcurrency"?: Org_OData_Core_V1.OptimisticConcurrency;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"Ordered"?: Org_OData_Core_V1.Ordered;
	"PositionalInsert"?: Org_OData_Core_V1.PositionalInsert;
	"AlternateKeys"?: Org_OData_Core_V1.AlternateKeys;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ResourcePath"?: Org_OData_Core_V1.ResourcePath;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ResourcePath"?: Org_OData_Core_V1.ResourcePath;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ResourcePath"?: Org_OData_Core_V1.ResourcePath;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"OperationAvailable"?: Org_OData_Core_V1.OperationAvailable;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"Permissions"?: Org_OData_Core_V1.Permissions;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"OperationAvailable"?: Org_OData_Core_V1.OperationAvailable;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"DefaultNamespace"?: Org_OData_Core_V1.DefaultNamespace;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Core {
	"Revisions"?: Org_OData_Core_V1.Revisions;
	"Description"?: Org_OData_Core_V1.Description;
	"LongDescription"?: Org_OData_Core_V1.LongDescription;
	"Links"?: Org_OData_Core_V1.Links;
	"Messages"?: Org_OData_Core_V1.Messages;
	"ValueException"?: Org_OData_Core_V1.ValueException;
	"ResourceException"?: Org_OData_Core_V1.ResourceException;
	"DataModificationException"?: Org_OData_Core_V1.DataModificationException;
	"ContentID"?: Org_OData_Core_V1.ContentID;
	"MayImplement"?: Org_OData_Core_V1.MayImplement;
	"SymbolicName"?: Org_OData_Core_V1.SymbolicName;
}
