// @sap-ux/vocabularies-types/types/generated/Hierarchy_Edm.ts
import * as com_sap_vocabularies_Hierarchy_v1 from "./Hierarchy";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Hierarchy {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Hierarchy {
	"RecursiveHierarchy"?: com_sap_vocabularies_Hierarchy_v1.RecursiveHierarchy;
}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Hierarchy {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Hierarchy {
	"RecursiveHierarchy"?: com_sap_vocabularies_Hierarchy_v1.RecursiveHierarchy;
}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Hierarchy {}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Hierarchy {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Hierarchy {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Hierarchy {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Hierarchy {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Hierarchy {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Hierarchy {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Hierarchy {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Hierarchy {}
