// @sap-ux/vocabularies-types/types/generated/HTML5_Edm.ts
import * as com_sap_vocabularies_HTML5_v1 from "./HTML5";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_HTML5 {}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_HTML5 {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_HTML5 {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_HTML5 {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_HTML5 {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_HTML5 {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_HTML5 {}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_HTML5 {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_HTML5 {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_HTML5 {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_HTML5 {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_HTML5 {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_HTML5 {}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_HTML5 {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_HTML5 {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_HTML5 {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_HTML5 {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_HTML5 {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_HTML5 {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_HTML5 {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_HTML5 {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_HTML5 {
	"CssDefaults"?: com_sap_vocabularies_HTML5_v1.CssDefaults;
}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_HTML5 {}
