// @sap-ux/vocabularies-types/types/generated/Measures_Edm.ts
import * as Org_OData_Measures_V1 from "./Measures";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Measures {}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Measures {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Measures {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Measures {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Measures {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Measures {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Measures {
	"ISOCurrency"?: Org_OData_Measures_V1.ISOCurrency;
	"Scale"?: Org_OData_Measures_V1.Scale;
	"Unit"?: Org_OData_Measures_V1.Unit;
	"DurationGranularity"?: Org_OData_Measures_V1.DurationGranularity;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Measures {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Measures {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Measures {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Measures {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Measures {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Measures {}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Measures {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Measures {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Measures {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Measures {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Measures {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Measures {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Measures {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Measures {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Measures {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Measures {}
