// @sap-ux/vocabularies-types/types/generated/PersonalData.ts
import * as Core from "./Core";
import * as Validation from "./Validation";
import * as Common from "./Common";
import * as Edm from "../Edm";
import AnnotationTerm = Edm.AnnotationTerm;
import PropertyAnnotationValue = Edm.PropertyAnnotationValue;
import EnumValue = Edm.EnumValue;
import ComplexType = Edm.ComplexType;

// Term
/**
  Primary meaning of the entities in the annotated entity set
*/
export type EntitySemantics = AnnotationTerm<PropertyAnnotationValue<EntitySemanticsType>>;

// Term
/**
  Role of the data subjects in this set (e.g. employee, customer)
*/
export type DataSubjectRole = AnnotationTerm<PropertyAnnotationValue<Edm.String>>;

// Term
/**
  Language-dependent description of the role of the data subjects in this set (e.g. employee, customer)
*/
export type DataSubjectRoleDescription = AnnotationTerm<PropertyAnnotationValue<Edm.String>>;

// Typedefinition
/**
  Primary meaning of the data contained in the annotated entity set
*/
export type EntitySemanticsType = Edm.String;

// Term
/**
  Primary meaning of the personal data contained in the annotated property
*/
export type FieldSemantics = AnnotationTerm<PropertyAnnotationValue<FieldSemanticsType>>;

// Typedefinition
/**
  Primary meaning of a data field
*/
export type FieldSemanticsType = Edm.String;

// Term
/**
  Property contains potentially personal data
*/
export type IsPotentiallyPersonal = AnnotationTerm<PropertyAnnotationValue<Core.Tag>>;

// Term
/**
  Property contains potentially sensitive personal data
*/
export type IsPotentiallySensitive = AnnotationTerm<PropertyAnnotationValue<Core.Tag>>;

// Term
/**
  Property contains a user id
*/
export type IsUserID = AnnotationTerm<PropertyAnnotationValue<Core.Tag>>;
export const enum PersonalDataAnnotationTerms {
	EntitySemantics = "com.sap.vocabularies.PersonalData.v1.EntitySemantics",
	DataSubjectRole = "com.sap.vocabularies.PersonalData.v1.DataSubjectRole",
	DataSubjectRoleDescription = "com.sap.vocabularies.PersonalData.v1.DataSubjectRoleDescription",
	FieldSemantics = "com.sap.vocabularies.PersonalData.v1.FieldSemantics",
	IsPotentiallyPersonal = "com.sap.vocabularies.PersonalData.v1.IsPotentiallyPersonal",
	IsPotentiallySensitive = "com.sap.vocabularies.PersonalData.v1.IsPotentiallySensitive",
	IsUserID = "com.sap.vocabularies.PersonalData.v1.IsUserID"
}
export const enum PersonalDataAnnotationTypes {}
