// @sap-ux/vocabularies-types/types/generated/PersonalData_Edm.ts
import * as com_sap_vocabularies_PersonalData_v1 from "./PersonalData";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_PersonalData {}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_PersonalData {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_PersonalData {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_PersonalData {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_PersonalData {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_PersonalData {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_PersonalData {
	"FieldSemantics"?: com_sap_vocabularies_PersonalData_v1.FieldSemantics;
	"IsPotentiallyPersonal"?: com_sap_vocabularies_PersonalData_v1.IsPotentiallyPersonal;
	"IsPotentiallySensitive"?: com_sap_vocabularies_PersonalData_v1.IsPotentiallySensitive;
	"IsUserID"?: com_sap_vocabularies_PersonalData_v1.IsUserID;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_PersonalData {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_PersonalData {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_PersonalData {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_PersonalData {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_PersonalData {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_PersonalData {
	"EntitySemantics"?: com_sap_vocabularies_PersonalData_v1.EntitySemantics;
	"DataSubjectRole"?: com_sap_vocabularies_PersonalData_v1.DataSubjectRole;
	"DataSubjectRoleDescription"?: com_sap_vocabularies_PersonalData_v1.DataSubjectRoleDescription;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_PersonalData {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_PersonalData {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_PersonalData {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_PersonalData {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_PersonalData {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_PersonalData {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_PersonalData {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_PersonalData {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_PersonalData {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_PersonalData {}
