// @sap-ux/vocabularies-types/types/generated/Session_Edm.ts
import * as com_sap_vocabularies_Session_v1 from "./Session";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Session {}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Session {}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Session {}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Session {}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Session {}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Session {}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Session {}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Session {}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Session {}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Session {}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Session {}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Session {}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Session {
	"StickySessionSupported"?: com_sap_vocabularies_Session_v1.StickySessionSupported;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Session {}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Session {}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Session {}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Session {}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Session {}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Session {}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Session {}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Session {}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Session {}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Session {}
