// @sap-ux/vocabularies-types/types/generated/UI_Edm.ts
import * as com_sap_vocabularies_UI_v1 from "./UI";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_UI {
	"HeaderInfo"?: com_sap_vocabularies_UI_v1.HeaderInfo;
	"Identification"?: com_sap_vocabularies_UI_v1.Identification;
	"Badge"?: com_sap_vocabularies_UI_v1.Badge;
	"LineItem"?: com_sap_vocabularies_UI_v1.LineItem;
	"StatusInfo"?: com_sap_vocabularies_UI_v1.StatusInfo;
	"FieldGroup"?: com_sap_vocabularies_UI_v1.FieldGroup;
	"ConnectedFields"?: com_sap_vocabularies_UI_v1.ConnectedFields;
	"GeoLocations"?: com_sap_vocabularies_UI_v1.GeoLocations;
	"GeoLocation"?: com_sap_vocabularies_UI_v1.GeoLocation;
	"Contacts"?: com_sap_vocabularies_UI_v1.Contacts;
	"MediaResource"?: com_sap_vocabularies_UI_v1.MediaResource;
	"DataPoint"?: com_sap_vocabularies_UI_v1.DataPoint;
	"KPI"?: com_sap_vocabularies_UI_v1.KPI;
	"Chart"?: com_sap_vocabularies_UI_v1.Chart;
	"CriticalityLabels"?: com_sap_vocabularies_UI_v1.CriticalityLabels;
	"SelectionFields"?: com_sap_vocabularies_UI_v1.SelectionFields;
	"Facets"?: com_sap_vocabularies_UI_v1.Facets;
	"HeaderFacets"?: com_sap_vocabularies_UI_v1.HeaderFacets;
	"QuickViewFacets"?: com_sap_vocabularies_UI_v1.QuickViewFacets;
	"QuickCreateFacets"?: com_sap_vocabularies_UI_v1.QuickCreateFacets;
	"FilterFacets"?: com_sap_vocabularies_UI_v1.FilterFacets;
	"SelectionPresentationVariant"?: com_sap_vocabularies_UI_v1.SelectionPresentationVariant;
	"PresentationVariant"?: com_sap_vocabularies_UI_v1.PresentationVariant;
	"SelectionVariant"?: com_sap_vocabularies_UI_v1.SelectionVariant;
	"TextArrangement"?: com_sap_vocabularies_UI_v1.TextArrangement;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_UI {
	"ValueCriticality"?: com_sap_vocabularies_UI_v1.ValueCriticality;
	"CriticalityLabels"?: com_sap_vocabularies_UI_v1.CriticalityLabels;
	"IsImageURL"?: com_sap_vocabularies_UI_v1.IsImageURL;
	"IsImage"?: com_sap_vocabularies_UI_v1.IsImage;
	"MultiLineText"?: com_sap_vocabularies_UI_v1.MultiLineText;
	"Placeholder"?: com_sap_vocabularies_UI_v1.Placeholder;
	"Hidden"?: com_sap_vocabularies_UI_v1.Hidden;
	"HiddenFilter"?: com_sap_vocabularies_UI_v1.HiddenFilter;
	"DataFieldDefault"?: com_sap_vocabularies_UI_v1.DataFieldDefault;
	"OrderBy"?: com_sap_vocabularies_UI_v1.OrderBy;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
	"RecommendationList"?: com_sap_vocabularies_UI_v1.RecommendationList;
	"ExcludeFromNavigationContext"?: com_sap_vocabularies_UI_v1.ExcludeFromNavigationContext;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_UI {
	"Hidden"?: com_sap_vocabularies_UI_v1.Hidden;
	"HiddenFilter"?: com_sap_vocabularies_UI_v1.HiddenFilter;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_UI {
	"ValueCriticality"?: com_sap_vocabularies_UI_v1.ValueCriticality;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_UI {
	"ThingPerspective"?: com_sap_vocabularies_UI_v1.ThingPerspective;
	"IsImageURL"?: com_sap_vocabularies_UI_v1.IsImageURL;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_UI {
	"ParameterDefaultValue"?: com_sap_vocabularies_UI_v1.ParameterDefaultValue;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
	"RecommendationList"?: com_sap_vocabularies_UI_v1.RecommendationList;
}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_UI {
	"KPI"?: com_sap_vocabularies_UI_v1.KPI;
	"SelectionPresentationVariant"?: com_sap_vocabularies_UI_v1.SelectionPresentationVariant;
	"PresentationVariant"?: com_sap_vocabularies_UI_v1.PresentationVariant;
	"SelectionVariant"?: com_sap_vocabularies_UI_v1.SelectionVariant;
	"CreateHidden"?: com_sap_vocabularies_UI_v1.CreateHidden;
	"UpdateHidden"?: com_sap_vocabularies_UI_v1.UpdateHidden;
	"DeleteHidden"?: com_sap_vocabularies_UI_v1.DeleteHidden;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_UI {
	"FieldGroup"?: com_sap_vocabularies_UI_v1.FieldGroup;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_UI {
	"FieldGroup"?: com_sap_vocabularies_UI_v1.FieldGroup;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_UI {
	"FieldGroup"?: com_sap_vocabularies_UI_v1.FieldGroup;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_UI {
	"TextArrangement"?: com_sap_vocabularies_UI_v1.TextArrangement;
	"Importance"?: com_sap_vocabularies_UI_v1.Importance;
	"Criticality"?: com_sap_vocabularies_UI_v1.Criticality;
	"CriticalityCalculation"?: com_sap_vocabularies_UI_v1.CriticalityCalculation;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_UI {
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_UI {
	"IsSummary"?: com_sap_vocabularies_UI_v1.IsSummary;
	"PartOfPreview"?: com_sap_vocabularies_UI_v1.PartOfPreview;
	"Map"?: com_sap_vocabularies_UI_v1.Map;
	"Gallery"?: com_sap_vocabularies_UI_v1.Gallery;
	"Importance"?: com_sap_vocabularies_UI_v1.Importance;
	"Hidden"?: com_sap_vocabularies_UI_v1.Hidden;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_UI {
	"MultiLineText"?: com_sap_vocabularies_UI_v1.MultiLineText;
	"RecommendationState"?: com_sap_vocabularies_UI_v1.RecommendationState;
}
