// @sap-ux/vocabularies-types/types/generated/Validation_Edm.ts
import * as Org_OData_Validation_V1 from "./Validation";

// Type containing all possible annotations to use for EntityContainer
export interface EntityContainerAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Schema
export interface SchemaAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Reference
export interface ReferenceAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for EntityType
export interface EntityTypeAnnotations_Validation {
	"Constraint"?: Org_OData_Validation_V1.Constraint;
	"ItemsOf"?: Org_OData_Validation_V1.ItemsOf;
	"OpenPropertyTypeConstraint"?: Org_OData_Validation_V1.OpenPropertyTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for EnumType
export interface EnumTypeAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for ComplexType
export interface ComplexTypeAnnotations_Validation {
	"Constraint"?: Org_OData_Validation_V1.Constraint;
	"ItemsOf"?: Org_OData_Validation_V1.ItemsOf;
	"OpenPropertyTypeConstraint"?: Org_OData_Validation_V1.OpenPropertyTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Property
export interface PropertyAnnotations_Validation {
	"Pattern"?: Org_OData_Validation_V1.Pattern;
	"Minimum"?: Org_OData_Validation_V1.Minimum;
	"Maximum"?: Org_OData_Validation_V1.Maximum;
	"AllowedValues"?: Org_OData_Validation_V1.AllowedValues;
	"MultipleOf"?: Org_OData_Validation_V1.MultipleOf;
	"Constraint"?: Org_OData_Validation_V1.Constraint;
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"AllowedTerms"?: Org_OData_Validation_V1.AllowedTerms;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for NavigationProperty
export interface NavigationPropertyAnnotations_Validation {
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for TypeDefinition
export interface TypeDefinitionAnnotations_Validation {
	"AllowedValues"?: Org_OData_Validation_V1.AllowedValues;
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Term
export interface TermAnnotations_Validation {
	"Pattern"?: Org_OData_Validation_V1.Pattern;
	"Minimum"?: Org_OData_Validation_V1.Minimum;
	"Maximum"?: Org_OData_Validation_V1.Maximum;
	"MultipleOf"?: Org_OData_Validation_V1.MultipleOf;
	"AllowedTerms"?: Org_OData_Validation_V1.AllowedTerms;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Parameter
export interface ParameterAnnotations_Validation {
	"Pattern"?: Org_OData_Validation_V1.Pattern;
	"Minimum"?: Org_OData_Validation_V1.Minimum;
	"Maximum"?: Org_OData_Validation_V1.Maximum;
	"AllowedValues"?: Org_OData_Validation_V1.AllowedValues;
	"MultipleOf"?: Org_OData_Validation_V1.MultipleOf;
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for ReturnType
export interface ReturnTypeAnnotations_Validation {
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for EntitySet
export interface EntitySetAnnotations_Validation {
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Singleton
export interface SingletonAnnotations_Validation {
	"DerivedTypeConstraint"?: Org_OData_Validation_V1.DerivedTypeConstraint;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for ActionImport
export interface ActionImportAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for FunctionImport
export interface FunctionImportAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Action
export interface ActionAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Function
export interface FunctionAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Include
export interface IncludeAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Annotation
export interface AnnotationAnnotations_Validation {
	"Exclusive"?: Org_OData_Validation_V1.Exclusive;
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for Collection
export interface CollectionAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
	"MaxItems"?: Org_OData_Validation_V1.MaxItems;
	"MinItems"?: Org_OData_Validation_V1.MinItems;
}

// Type containing all possible annotations to use for Record
export interface RecordAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}

// Type containing all possible annotations to use for PropertyValue
export interface PropertyValueAnnotations_Validation {
	"ApplicableTerms"?: Org_OData_Validation_V1.ApplicableTerms;
}
