// @sap/ux-specification/dist/src/sync/common/flexUtils.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const flexUtils = __importStar(require("@ui5/flexibility-utils"));
const changesUtils = flexUtils.default.change;
const defaultValues = {
    isCustomer: false,
    sapui5Version: ' ',
    creatingTool: 'SAPFioriTools',
    type: 'propertyChange'
};
/**
 * Create a change as string (calls changeUtils from ui5-flexibility-utils)
 * @param change - properties required to create the change
 * @param manifest - the manifest.json file
 */
function createChangeString(change, manifest) {
    let changeString = '';
    try {
        for (const defaultValue in defaultValues) {
            if (change[defaultValue] === undefined) {
                change[defaultValue] = defaultValues[defaultValue];
            }
        }
        changeString = changesUtils.createChangeString(change, manifest);
    }
    catch (error) {
        console.error(`Error while create flex change.`, error);
    }
    return changeString;
}
exports.createChangeString = createChangeString;
/**
 * Parse string and return the change as object (calls changeUtils from ui5-flexibility-utils)
 * @param changeString - string with serialized change
 */
function parseChangeString(changeString) {
    const change = changesUtils.parse(changeString);
    return change;
}
exports.parseChangeString = parseChangeString;
