// @sap/ux-specification/dist/src/sync/common/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExportArtifacts;
(function (ExportArtifacts) {
    ExportArtifacts["flex"] = "flex";
    ExportArtifacts["manifest"] = "manifest";
})(ExportArtifacts = exports.ExportArtifacts || (exports.ExportArtifacts = {}));
var DirName;
(function (DirName) {
    DirName["Schemas"] = ".schemas";
    DirName["Pages"] = "pages";
})(DirName = exports.DirName || (exports.DirName = {}));
var FileName;
(function (FileName) {
    FileName["App"] = "app.json";
})(FileName = exports.FileName || (exports.FileName = {}));
exports.FacetTitlePrefix = 'Facet ID: ';
var ControlType;
(function (ControlType) {
    ControlType["Table"] = "sap.m.Table";
    ControlType["TableColumn"] = "sap.m.Column";
    ControlType["SmartTable"] = "sap.ui.comp.smarttable.SmartTable";
    ControlType["SmartFilterBar"] = "sap.ui.comp.smartfilterbar.SmartFilterBar";
    ControlType["SmartChart"] = "sap.ui.comp.smartchart.SmartChart";
    ControlType["Button"] = "sap.m.Button";
    ControlType["ToolbarButton"] = "sap.m.OverflowToolbarButton";
    ControlType["Avatar"] = "sap.f.Avatar";
    ControlType["ObjectPageDynamicHeaderTitle"] = "sap.uxap.ObjectPageDynamicHeaderTitle";
    ControlType["ObjectPageGridProperties"] = "sap.ui.layout.GridData";
    ControlType["ObjectPageHeader"] = "sap.uxap.ObjectPageHeader";
    ControlType["ObjectPageLayout"] = "sap.uxap.ObjectPageLayout";
    ControlType["DynamicPage"] = "sap.f.DynamicPage";
    ControlType["Form"] = "sap.ui.layout.form";
    ControlType["Chart"] = "sap.suite.ui.microchart";
})(ControlType = exports.ControlType || (exports.ControlType = {}));
exports.BindingPropertyRegexAsString = "^{[A-Za-z0-9{}&$@? _|,>'[\\]\\/:=.]+}$";
