// @sap/ux-specification/dist/src/specification/v2/controls/Card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChartCardType;
(function (ChartCardType) {
    ChartCardType["cardBubble"] = "cardBubble";
    ChartCardType["cardchartsline"] = "cardchartsline";
    ChartCardType["cardchartsdonut"] = "cardchartsdonut";
})(ChartCardType = exports.ChartCardType || (exports.ChartCardType = {}));
var CardTemplateType;
(function (CardTemplateType) {
    CardTemplateType["analytical"] = "sap.ovp.cards.charts.analytical";
    CardTemplateType["list"] = "sap.ovp.cards.list";
    CardTemplateType["linklist"] = "sap.ovp.cards.linklist";
    CardTemplateType["table"] = "sap.ovp.cards.table";
    CardTemplateType["stack"] = "sap.ovp.cards.stack";
})(CardTemplateType = exports.CardTemplateType || (exports.CardTemplateType = {}));
var CardSettingsType;
(function (CardSettingsType) {
    CardSettingsType["analyticalCardSettings"] = "analyticalCardSettings";
    CardSettingsType["listCardSettings"] = "listCardSettings";
    CardSettingsType["stackCardSettings"] = "stackCardSettings";
    CardSettingsType["linkListCardSettings"] = "linkListCardSettings";
    CardSettingsType["tableCardSettings"] = "tableCardSettings";
})(CardSettingsType = exports.CardSettingsType || (exports.CardSettingsType = {}));
var ListFlavorType;
(function (ListFlavorType) {
    ListFlavorType["standard"] = "standard";
    ListFlavorType["bar"] = "bar";
    ListFlavorType["carousel"] = "carousel";
})(ListFlavorType = exports.ListFlavorType || (exports.ListFlavorType = {}));
var ListTypeType;
(function (ListTypeType) {
    ListTypeType["extended"] = "extended";
    ListTypeType["condensed"] = "condensed";
})(ListTypeType = exports.ListTypeType || (exports.ListTypeType = {}));
var SortOrderType;
(function (SortOrderType) {
    SortOrderType["ascending"] = "ascending";
    SortOrderType["descending"] = "descending";
})(SortOrderType = exports.SortOrderType || (exports.SortOrderType = {}));
var LinkListFlavorType;
(function (LinkListFlavorType) {
    LinkListFlavorType["standard"] = "standard";
    LinkListFlavorType["carousel"] = "carousel";
})(LinkListFlavorType = exports.LinkListFlavorType || (exports.LinkListFlavorType = {}));
