// @sap/ux-specification/dist/src/specification/v2/controls/Table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HAlign;
(function (HAlign) {
    HAlign["initial"] = "Initial";
    HAlign["left"] = "Left";
    HAlign["center"] = "Center";
    HAlign["right"] = "Right";
    HAlign["begin"] = "Begin";
    HAlign["end"] = "End";
})(HAlign = exports.HAlign || (exports.HAlign = {}));
var TableTypeV2;
(function (TableTypeV2) {
    TableTypeV2["ResponsiveTable"] = "ResponsiveTable";
    TableTypeV2["GridTable"] = "GridTable";
    TableTypeV2["AnalyticalTable"] = "AnalyticalTable";
    TableTypeV2["TreeTable"] = "TreeTable";
})(TableTypeV2 = exports.TableTypeV2 || (exports.TableTypeV2 = {}));
var PopinLayout;
(function (PopinLayout) {
    PopinLayout["Block"] = "Block";
    PopinLayout["GridLarge"] = "GridLarge";
    PopinLayout["GridSmall"] = "GridSmall";
})(PopinLayout = exports.PopinLayout || (exports.PopinLayout = {}));
var DefaultContentView;
(function (DefaultContentView) {
    DefaultContentView["charttable"] = "charttable";
    DefaultContentView["chart"] = "chart";
    DefaultContentView["table"] = "table";
})(DefaultContentView = exports.DefaultContentView || (exports.DefaultContentView = {}));
var DefaultFilterMode;
(function (DefaultFilterMode) {
    DefaultFilterMode["visual"] = "visual";
    DefaultFilterMode["compact"] = "compact";
})(DefaultFilterMode = exports.DefaultFilterMode || (exports.DefaultFilterMode = {}));
var LoadDataOnAppLaunchSettings;
(function (LoadDataOnAppLaunchSettings) {
    LoadDataOnAppLaunchSettings["always"] = "always";
    LoadDataOnAppLaunchSettings["never"] = "never";
    LoadDataOnAppLaunchSettings["ifAnyFilterExist"] = "ifAnyFilterExist";
})(LoadDataOnAppLaunchSettings = exports.LoadDataOnAppLaunchSettings || (exports.LoadDataOnAppLaunchSettings = {}));
